### ISSUE TITLE
[Feature Request]: Reflection Padding for VAE

### ISSUE THREAD

**Issue by Anzhc** [Panchovix/stable-diffusion-webui-reForge#422]
### Is there an existing issue for this?

- [x] I have searched the existing issues and checked the recent builds/commits

### What would your feature do ?

Hey @Panchovix , it's me. Saw that you're working on adding reflection padding, and thought i'd share how i hardcoded it on my end in reforge (and it works). It was a bit different from comfy, so might help, as i saw you're not sure if it works(And i think it's too dangerous to update rn to test it xD).

### Proposed workflow

Here is how i did it:
modules/sd.py i use this function to pad:
```
    def _use_reflection_padding(self):
        """Set all Conv2d layers with padding to use reflection padding."""
        for module in self.first_stage_model.modules():
            from torch import nn
            logging.info(self)
            if isinstance(module, nn.Conv2d):
                pad_h, pad_w = module.padding if isinstance(module.padding, tuple) else (module.padding, module.padding)
                if pad_h > 0 or pad_w > 0:
                    module.padding_mode = "reflect"
        logging.info("Setting reflective padding")
```
Then i call
`self._use_reflection_padding()`
At the end of VAE creation:
```
class VAE:
    def __init__(self, sd=None, device=None, config=None, dtype=None, metadata=None, no_init=False):
...
...
...
        else:
            self.first_stage_model = AutoencoderKL(**(config['params']))
        self.first_stage_model = self.first_stage_model.eval()

        m, u = self.first_stage_model.load_state_dict(sd, strict=False)
        if len(m) > 0:
            logging.warning("Missing VAE keys {}".format(m))

        if len(u) > 0:
            logging.debug("Leftover VAE keys {}".format(u))

        if device is None:
            device = model_management.vae_device()
        self.device = device
        offload_device = model_management.vae_offload_device()
        if dtype is None:
            dtype = model_management.vae_dtype(self.device, self.working_dtypes)
        self.vae_dtype = dtype
        self.first_stage_model.to(self.vae_dtype)
        self.output_device = model_management.intermediate_device()

        self.patcher = ldm_patched.modules.model_patcher.ModelPatcher(self.first_stage_model, load_device=self.device, offload_device=offload_device)
        logging.info("VAE load device: {}, offload device: {}, dtype: {}".format(self.device, offload_device, self.vae_dtype))

        self._use_reflection_padding()
```

That is all that need to make it work, hopefully you can make it in more editable way.

**Comment by Panchovix** [comment]
Hello there, thanks! Wondering, did you add the function inside the VAE class?

I.e. like

```
        self.first_stage_model = self.first_stage_model.eval()

        m, u = self.first_stage_model.load_state_dict(sd, strict=False)
        if len(m) > 0:
            logging.warning("Missing VAE keys {}".format(m))

        if len(u) > 0:
            logging.debug("Leftover VAE keys {}".format(u))

    def _use_reflection_padding(self):
        """Set all Conv2d layers with padding to use reflection padding."""
        for module in self.first_stage_model.modules():
            from torch import nn
            logging.info(self)
            if isinstance(module, nn.Conv2d):
                pad_h, pad_w = module.padding if isinstance(module.padding, tuple) else (module.padding, module.padding)
                if pad_h > 0 or pad_w > 0:
                    module.padding_mode = "reflect"
        logging.info("Setting reflective padding")

        if device is None:
            device = model_management.vae_device()
        self.device = device
        offload_device = model_management.vae_offload_device()
        if dtype is None:
            dtype = model_management.vae_dtype(self.device, self.working_dtypes)
        self.vae_dtype = dtype
        self.first_stage_model.to(self.vae_dtype)
        self.output_device = model_management.intermediate_device()

        self.patcher = ldm_patched.modules.model_patcher.ModelPatcher(self.first_stage_model, load_device=self.device, offload_device=offload_device)
        logging.info("VAE load device: {}, offload device: {}, dtype: {}".format(self.device, offload_device, self.vae_dtype))

        self._use_reflection_padding()
```

I think your implementation makes more sense if it doesn't changes default VAE behavior (which I think is the case here so that's good)

**Comment by Anzhc** [comment]
I called it hardcoded for a reason. It does change behaviour of all VAEs, if they have conv layers. So for example default SDXL VAE would also get this treatment, but it is not trained with reflection padding, so adding that to it would cause artifacts on edges.

Your approach with extension is most reasonable one, i just wanted to share a code example that does successfully patch VAEs, so if you could somehow make it in a checkbox extension - that would be best.

What you currently did is exactly what i did(so code looks correct), so right now VAEs trained with reflection padding are patched, but all other VAEs will break a bit.

**Comment by Panchovix** [comment]
Also pardon my ignorance, but are results like this normal with a model without reflection padding, but using the reflection padding VAE?

<img width="896" height="1088" alt="Image" src="https://github.com/user-attachments/assets/a9794cdf-95db-4099-9e1a-56a6814f5a9a" />

```
Steps: 25, Sampler: DPM++ SDE, Schedule type: SGM Uniform, CFG scale: 5, Seed: 138344579, Size: 896x1088, Model hash: 2a7fc06348, Model: naiXLVpred102d_colorized, VAE hash: e3488c254f, VAE: MS-LC-EQ-D-VR VAE B3.safetensors, Emphasis: No norm, Version: f1.0.0v2-v1.10.1RC-latest-2450-g618e02a2
```

**Comment by Anzhc** [comment]
Awesome. That'll make it much comfier now.

<img width="63" height="51" alt="Image" src="https://github.com/user-attachments/assets/0b602590-de98-47aa-937f-8e3c9cfdc72f" />

What you see in your image is not related to reflection - that's an effect of EQ VAE on non-eq model, that's expected.

For testing if padding works or not you can try any vae on any model, but use some plain background, like red background, and look at edges. You should see changes. If normal VAE looks normal with option off and gets some artifacts with it on and vice versa for eq vae - it works.

**Comment by Anzhc** [comment]
> [@Anzhc](https://github.com/Anzhc) thanks for the headup, added a new option to set enable or disable this. Disabled by default.
> 
> Tried to replicate the commit on the forge2 based one but it doesn't seem to work, I guess it is different at some place [df1d594](https://github.com/Panchovix/stable-diffusion-webui-reForge/commit/df1d5948f923c227902362f311512a0d26f5ac61)

And about comfy way of implementing it - i find that some functions from comfy backend just refuse to work.
For example i tried to use flow matching in advanced model sampling some time ago(using SD3 sampling, same as in comfy)(To test SDXL finetune(it works in comfy)), no matter what i did it wasn't working no matter what i tried, despite code being identical between comfy and reforge in this case.

Couldn't find what the issue was either, some things are just weird.

<img width="63" height="51" alt="Image" src="https://github.com/user-attachments/assets/cc16bc30-d40a-4ef7-a4e0-426ecae459f7" />

**Comment by Anzhc** [comment]
@Panchovix I verified, it is working.

You were not able to verify it because switch owrks only at app startup, but has no effect before hard restart.

<img width="1420" height="1135" alt="Image" src="https://github.com/user-attachments/assets/af81fe6a-be9a-4b21-b15a-d634ecc203d4" />
It does apply padding if you start webui with option on, and does not if it was off. Already much better than my hardcoding though :D

**Comment by Anzhc** [comment]
Also switching vaes back and forth after changing option does not apply/remove padding, so it is likely that VAE swapping code does not trigger the class code, somehow.
